src/register/mcounteren.rs

```diff
@@ -0,0 +1,494 @@
+//! mcounteren register
+
+
+use bit_field::BitField;
+//use core::mem::size_of;
+
+/// mcounteren register
+#[derive(Clone, Copy, Debug)]
+pub struct Mcounteren {
+    bits: usize,
+}
+
+
+/// Enable/disable in mcounteren bits
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum State {
+    Enabled = 1,
+    Disabled = 0,
+}
+
+/// Enable/disable in mcounteren bits
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub enum Index {
+    Hpm3 = 3,
+    Hpm4 = 4,
+    Hpm5 = 5,
+    Hpm6 = 6,
+    Hpm7 = 7,
+    Hpm8 = 8,
+    Hpm9 = 9,
+    Hpm10 = 10,
+    Hpm11 = 11,
+    Hpm12 = 12,
+    Hpm13 = 13,
+    Hpm14 = 14,
+    Hpm15 = 15,
+    Hpm16 = 16,
+    Hpm17 = 17,
+    Hpm18 = 18,
+    Hpm19 = 19,
+    Hpm20 = 20,
+    Hpm21 = 21,
+    Hpm22 = 22,
+    Hpm23 = 23,
+    Hpm24 = 24,
+    Hpm25 = 25,
+    Hpm26 = 26,
+    Hpm27 = 27,
+    Hpm28 = 28,
+    Hpm29 = 29,
+    Hpm30 = 30,
+    Hpm31 = 31,
+}
+
+
+impl Mcounteren {
+    /// User "cycle[h]" Enable
+    #[inline]
+    pub fn cy(&self) -> State {
+        match self.bits.get_bit(0) {
+            true => State::Enabled,
+            false => State::Disabled,
+        }
+    }
+
+    /// User "time[h]" Enable
+    #[inline]
+    pub fn tm(&self) -> State {
+        match self.bits.get_bit(1) {
+            true => State::Enabled,
+            false => State::Disabled,
+        }
+    }
+
+    /// User "instret[h]" Enable
+    #[inline]
+    pub fn ir(&self) -> State {
+        match self.bits.get_bit(2) {
+            true => State::Enabled,
+            false => State::Disabled,
+        }
+    }
+
+    /// User "hpm[x]" Enable (bits 3-31)
+    #[inline]
+    pub fn hpm(&self, index: Index) -> State {
```

Comment:
I'm not sure you need this `Index` enum here as it's quite trivial. It can be replaced with usize and `assert!` bounds check.